Makefile

```diff
@@ -84,25 +79,36 @@ fmt:
 	go fmt $(TEST_PKGS)
 
 container: hyperfed
-	cp -f $(HYPERFED_TARGET) images/federation-v2/hyperfed
-	$(DOCKER) build images/federation-v2 \
-		-t $(REGISTRY)/$(TARGET):$(GIT_VERSION)
+	cp -f $(HYPERFED_TARGET)-linux images/federation-v2/hyperfed
+	$(DOCKER) build images/federation-v2 -t $(IMAGE_NAME)
 	rm -f images/federation-v2/hyperfed
 
-$(BIN_DIR):
-	mkdir $(BIN_DIR)
+bindir:
+	mkdir -p $(BIN_DIR)
+
+COMMANDS := $(HYPERFED_TARGET) $(CONTROLLER_TARGET) $(KUBEFED2_TARGET)
+OSES := linux darwin
+ALL_BINS :=
+
+define OS_template
+$(1)-$(2): bindir
+	$(DOCKER_BUILD) '$(TARGET_ENV_$(2)) $(GO_BUILDCMD) -o $(1)-$(2) cmd/$(3)/main.go'
+ALL_BINS := $(ALL_BINS) $(1)-$(2)
+endef
+$(foreach cmd, $(COMMANDS), $(foreach os, $(OSES), $(eval $(call OS_template, $(cmd),$(os),$(notdir $(cmd))))))
```

Comment:
It does not make sense to build binaries for all OSes as a default as it is unnecessary compilation time overhead. So please only build for the HOST OS and should be the default.
For release build, only kubefed2 is the client side binary( so it makes sense to build & release kubefed2 for all OS). the container image will be based on Linux for now.

Reply:
it only builds for the host os if you do eg. "make controller"